galaxy_ng/app/api/ui/viewsets/collection.py

```diff
@@ -6,38 +7,77 @@
 from pulp_ansible.app import viewsets as pulp_ansible_viewsets
 from pulp_ansible.app.models import (
+    AnsibleCollectionDeprecated,
     AnsibleDistribution,
     CollectionVersion,
     Collection,
     CollectionRemote,
 )
 from pulp_ansible.app.models import CollectionImport as PulpCollectionImport
+from rest_framework import mixins
 from rest_framework.exceptions import NotFound
 from rest_framework.response import Response
+import semantic_version
 
 from galaxy_ng.app.api import base as api_base
 from galaxy_ng.app.access_control import access_policy
 from galaxy_ng.app.api.ui import serializers, versioning
 from galaxy_ng.app.api.v3.serializers.sync import CollectionRemoteSerializer
 
 
-class CollectionFilter(pulp_ansible_viewsets.CollectionVersionFilter):
+class CollectionByCollectionVersionFilter(pulp_ansible_viewsets.CollectionVersionFilter):
     """pulp_ansible CollectionVersion filter for Collection viewset."""
     versioning_class = versioning.UIVersioning
     keywords = filters.CharFilter(field_name="keywords", method="filter_by_q")
+    deprecated = filters.BooleanFilter()
 
 
-class CollectionViewSet(api_base.LocalSettingsMixin, pulp_ansible_galaxy_views.CollectionViewSet):
+class CollectionViewSet(
+    api_base.GenericViewSet,
+    mixins.ListModelMixin,
+    mixins.RetrieveModelMixin,
+    pulp_ansible_galaxy_views.AnsibleDistributionMixin,
+):
     """Viewset that uses CollectionVersion's within distribution to display data for Collection's.
 
-    Collection list is filterable by CollectionFilter and includes latest CollectionVersion.
+    Collection list is filterable by FilterSet and includes latest CollectionVersion.
 
     Collection detail includes CollectionVersion that is latest or via query param 'version'.
     """
     versioning_class = versioning.UIVersioning
-    filterset_class = CollectionFilter
+    filterset_class = CollectionByCollectionVersionFilter
     permission_classes = [access_policy.CollectionAccessPolicy]
 
+    def get_queryset(self):
+        """Returns a CollectionVersions queryset for specified distribution."""
+        distro_content = self.get_distro_content(self.kwargs["path"])
+        repo_version = self.get_repository_version(self.kwargs["path"])
+
+        versions = CollectionVersion.objects.filter(pk__in=distro_content).values_list(
+            "collection_id",
+            "version",
+        )
+
+        collection_versions = {}
+        for collection_id, version in versions:
+            value = collection_versions.get(str(collection_id))
+            if not value or semantic_version.Version(version) > semantic_version.Version(value):
+                collection_versions[str(collection_id)] = version
+
+        if not collection_versions.items():
+            return CollectionVersion.objects.none()
+
+        query_params = Q()
+        for collection_id, version in collection_versions.items():
+            query_params |= Q(collection_id=collection_id, version=version)
```

Comment:
Will this build up a where clause that includes a 'where collection_id=whatever and version=whatever' for every collection version in the distro?

Reply:
Doesn't seem to. Dumped a sql log of a GET /api/automation-hub/_ui/v1/collections 
at https://gist.github.com/alikins/02c1be49b812839551536186dfca4ea3